### PR TITLE
President rules outgoing

### PR DESCRIPTION
**Pull request by akowal3**
# Summary
Added transmission of rules & variables for tax/allocation
Added communication when no decision is made regarding tax/allocation
Introduced Tax/AllocationDecision struct to wrap all the necessary data into a single struct when sending it to clients

## Auxilary Info
The vast majority of additions is in tests

## Test Plan
Added tax/allocation tests. All tests are passing. IIGO runs

### CODE REVIEW

File: internal/common/baseclient/iigo.go [hunk 1]
```diff
@@ -52,8 +52,26 @@ func (c *BaseClient) TaxTaken(shared.Resources) {
 
 // GetTaxContribution gives value of how much the island wants to pay in taxes
 func (c *BaseClient) GetTaxContribution() shared.Resources {
-	// TODO: Implement common pool contribution greater than or equal to tax.
-	return 0
+	// If no new communication was received and there is no rule governing taxation, it will still recieve last valid tax
+	gameState := c.ServerReadHandle.GetGameState()
```
Comment (tobhil98) [inline internal/common/baseclient/iigo.go]: Maybe rename to clientGameState, so it avoids confusion with the gameState in IIGO?

File: internal/common/baseclient/iigo.go [hunk 2]
```diff
@@ -52,8 +52,26 @@ func (c *BaseClient) TaxTaken(shared.Resources) {
...
+
+	for i := range presidentCommunications {
+		msg := presidentCommunications[len(presidentCommunications)-1-i]
+		if msg[shared.Tax].T == shared.CommunicationTax {
+			newTaxDecision = msg[shared.Tax].TaxDecision
+			break
+		}
+	}
+
+	if newTaxDecision.TaxDecided {
```
Comment (tobhil98) [inline internal/common/baseclient/iigo.go:68]: Can we guarantee that the for loop will always initialise the newTaxDecision? You should really be initialising this to false inacse the for loop does not modify it.

File: internal/common/baseclient/iigo.go [hunk 3]
```diff
@@ -64,5 +82,22 @@ func (c *BaseClient) GetSanctionPayment() shared.Resources {
...
+	var newAllocationDecision shared.AllocationDecision
+	for i := range presidentCommunications {
+		msg := presidentCommunications[len(presidentCommunications)-1-i]
+		if msg[shared.Allocation].T == shared.CommunicationAllocation {
+			newAllocationDecision = msg[shared.Allocation].AllocationDecision
+			break
+		}
+	}
+
+	if newAllocationDecision.AllocationDecided {
```
Comment (tobhil98) [inline internal/common/baseclient/iigo.go:96]: Same here as above

File: internal/server/iigointernal/executive.go [hunk 4]
```diff
@@ -61,6 +69,10 @@ func (e *executive) setGameState(g *gamestate.GameState) {
 	e.gameState = g
 }
 
+func (e *executive) setGameConfig(c *config.IIGOConfig) {
```
Comment (tobhil98) [inline internal/server/iigointernal/executive.go]: Can you not use syncWithGame() instead?

File: internal/server/iigointernal/executive_test.go [hunk 5]
```diff
@@ -677,18 +684,82 @@ func TestReplyAllocationRequest(t *testing.T) {
 
 			setIIGOClients(&fakeClientMap)
 			tc.bPresident.setGameState(&fakeGameState)
+			tc.bPresident.setGameConfig(&fakeGameConfig)
```
Comment (tobhil98) [inline internal/server/iigointernal/executive_test.go]: See comment with syncWithGame()

**Review by tobhil98 — COMMENTED** [review]
Only minor comments

**Comment by akowal3** [comment]
incorporated feedback from @tobhil98 in the newest commit

**Review by tobhil98 — APPROVED** [review]
Looks good

**Pull request merged**